bugsnag-android-core/src/main/java/com/bugsnag/android/internal/DocumentPathDirective.kt

```diff
@@ -58,7 +61,7 @@ internal interface DocumentPathDirective<C> {
      */
     open class ListIndexDirective(private val index: Int) : DocumentPathDirective<MutableList<in Any>> {
         override fun newContainer(): MutableList<in Any> {
-            return mutableListOf()
+            return CopyOnWriteArrayList()
```

Comment:
This is the only way to be sure that `ConcurrentModificationException` never gets thrown. Alternatives would be to prohibit iteration over the document's contents at any level, or to include the caveat that anything iterating over the document in any way needs to also handle `ConcurrentModificationException`.

However, I figured that since there won't be more than 100 entries in a container at any level of the document ever, COW wouldn't be a big issue. Even if it were 1000 entries, that's only an 8k block of pointers to copy in two chunks for an insert or delete to that particular list/map in the document.

Also, not every journal entry is going to modify the same map or list. The most populous container we have is probably the stack trace, which is max 50 entries maybe, and more commonly 5-20. Next after that would be the breadcrumbs list, which we artificially limit.

Reply:
This is the only way to be sure that ConcurrentModificationException never gets thrown. Alternatives would be to prohibit iteration over the document's contents at any level, or to include the caveat that anything iterating over the document in any way needs to also handle ConcurrentModificationException.

However, I figured that since there won't be more than 100 entries in a container at any level of the document ever, COW wouldn't be a big issue. Even if it were 1000 entries, that's only an 8k block of pointers to copy in two chunks for an insert or delete to that particular list/map in the document.

Also, not every journal entry is going to modify the same map or list. The most populous container we have is probably the stack trace, which is max 50 entries maybe, and more commonly 5-20. Next after that would be the breadcrumbs list, which we artificially limit.

Reply:
I guess we can go with this for now and see later whether it impacts performance in real-world use.